File: ProblemSet/Week 3/Problem/JavaScript/LinkedList/src/auctionSystem.js
require('./blockchain');
const Auction = require('./auction');

class AuctionSystem {
    constructor() {
        this.auctions = [];
    }

    createAuction(item, startPrice, duration, description) {
        const auction = new Auction(item, startPrice, duration, description);
        this.auctions.push(auction);
    }

    placeBid(item, bidder, amount) {
        const auction = this.auctions.find((auction) => auction.item === item);
        if (auction && auction.isActive()) {
            auction.placeBid(bidder, amount);
            return true;
        }
        return false;
    }

    endAuction(item) {
        const auction = this.auctions.find((auction) => auction.item === item);
        if (auction && auction.isActive()) {
            const winner = auction.getWinner();
            if (winner) {
                const index = this.auctions.indexOf(auction);
                this.auctions.splice(index, 1); // Remove the auction from the list
                return winner;
            }
        }
        return null;
    }

    listBids() {
        const bids = [];
        this.auctions.forEach((auction) => {
            auction.bids.forEach((bid) => {
                bids.push({
                    item: auction.item,
                    bidder: bid.bidder,
                    amount: bid.amount,
                });
            });
        });
        return bids;
    }

    listAuctions() {
        return this.auctions.map((auction) => auction.item);
    }

    searchAuctions(keyword) {
        keyword.toLowerCase();
        return this.auctions.filter((auction) => {
            const lowerCaseItem = auction.item.toLowerCase();
            return lowerCaseItem.includes(lowerCaseItem);
        });
    }

    deleteAuction(item) {
        const index = this.auctions.findIndex((auction) => auction.item === item);
        if (index !== -1) {
            this.auctions.splice(index, 1);
            return true;
        }
        return false;
    }

    getAuctions() {
        return this.auctions;
    }

    setAuctions(auctions) {
        this.auctions = auctions;
    }
}

module.exports = AuctionSystem;

File: ProblemSet/Week 3/Problem/Python/LinkedList-v2-OOP/blockchain_auction_system/auction_system.py
from blockchain import Blockchain
from circular_linked_list import CircularLinkedList
from auction import Auction
import time

class AuctionSystem:
    """
    Manages the entire auction system.
    """
    def __init__(self):
        self.blockchain = Blockchain()
        self.auctions = CircularLinkedList()

    def create_auction(self, item, start_price, duration):
        """
        Create a new auction.
        """
        end_time = time.time() + duration
        new_auction = Auction(item, start_price, end_time)
        self.auctions.insert(new_auction)
        self.blockchain.add_block(f"New auction created: {item}")

    def place_bid(self, auction_item, bidder, amount):
        """
        Place a bid on an auction.
        """
        auction = self.auctions.search(lambda x: x.item == auction_item)
        if auction and auction.data.place_bid(bidder, amount):
            self.blockchain.add_block(f"Bid placed on {auction_item} by {bidder}: ${amount}")
            return True
        return False

    def end_auction(self, auction_item):
        """
        End an auction and determine the winner.
        """
        auction = self.auctions.search(lambda x: x.item == auction_item)
        if auction and time.time() >= auction.data.end_time:
            winner = auction.data.bids.tail.data if auction.data.bids.tail else None
            self.blockchain.add_block(f"Auction ended for {auction_item}. Winner: {winner[0] if winner else 'No winner'}")
            self.auctions.delete_node(auction.data)
            return winner
        return None